lib/jxl/encode.cc

```diff
@@ -1031,6 +1031,27 @@ JxlEncoderStatus JxlEncoderSetFrameDistance(
   return JXL_ENC_SUCCESS;
 }
 
+JxlEncoderStatus JxlEncoderSetExtraChannelDistance(
+    JxlEncoderFrameSettings* frame_settings, size_t index, float distance) {
+  if (index >= frame_settings->enc->metadata.m.num_extra_channels) {
+    return JXL_API_ERROR(frame_settings->enc, JXL_ENC_ERR_API_USAGE,
+                         "Invalid value for the index of extra channel");
+  }
+  if (distance < 0.f || distance > 25.f) {
+    return JXL_API_ERROR(frame_settings->enc, JXL_ENC_ERR_API_USAGE,
+                         "Distance has to be in [0.0..25.0] (corresponding to "
+                         "quality in [0.0..100.0])");
+  }
+  if (distance > 0.f && distance < 0.01f) {
+    distance = 0.01f;
+  }
+  frame_settings->values.cparams.ec_distance.resize(
```

Comment:
I don't like the resize happening here - I think we should maintain the invariant that cparams.ec_distanze.size() == number of extra channels.

Reply:
That would also make sense. The advantage of doing it this way is that setting color distance first and then setting alpha distance does the same thing as setting alpha distance first and then setting color distance, while not setting alpha distance always makes it use the color distance. If we want to maintain that invariant, then the ec_distance needs to get values before color distance is set, so then it would mean changing the color distance would need to overwrite the ec_distances (so not setting alpha distance makes it use the right default) so you would need to first set color distance and then alpha distance.

Then again this approach also isn't invariant to the order of setting things in the case that there are multiple extra channels and some of them get an explicit distance while others use the default. Maybe let's do something where distance -1 means "default", then we can have the invariant and also avoid requiring a specific order of setting things.

Reply:
Fully maintaining the invariant is annoying, because cparams doesn't know how many extra channels there are. I'll make it so that in our use of cparams in encode.cc, we do maintain the invariant that ec_distance.size() == nb_extra_chans, but in enc_modular.cc, if ec_distance.size() < nb_extra_chans, it just assumes the missing values are default (-1, same as color distance).

Reply:
PTAL if you like it better this way.